### PR TITLE
Create a unified set of flags for k8s clients

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Rename current Kubernetes options to legacy

In order to build a parallel set of options for use in the future, these
options will be renamed to a legacy version and migrate away from on
each individual component.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

Create a unified set of flags for k8s clients

Prow components will now expose three flags:

 --build-cluster : path to a Cluster file for build cluster credentials
 --kubeconfig : path to a kubeconfig for _all_ cluster credentials
 --context : context in the kubeconfig to use for the infra cluster

If these flags are not set, the default behavior is to use the
in-cluster context and client as the infrastrucutre cluster and the
build cluster.

If a set of build cluster credentials are passed in, they will add to
the build clusters available but we will still use the in-cluster
context for the infrastructure cluster.

If a kubeconfig is passed in, the default context will be used as the
infrastructure client context and the rest will be considered build
clusters.

If a kubeconfig and a context are passed in, the specified context will
be used as the infrastructure client and the others will be considered
build clusters.

This lets users opt into any additional behavior -- non-default
infrastructure clients or new build cluster clients -- as necessary.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

Implementation as discussed on Slack

/assign @cjwagner @fejta @munnerz 
/cc @sebastienvas @BenTheElder @krzyzacy

### CODE REVIEW

File: prow/flagutil/kubernetes_cluster_clients.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	// from resolution
+	resolved                   bool
+	dryRun                     bool
+	prowJobClientset           prow.Interface
+	kubernetesClientsByContext map[string]kubernetes.Interface
+}
+
+// AddFlags injects Kubernetes options into the given FlagSet.
+func (o *KubernetesOptions) AddFlags(fs *flag.FlagSet) {
+	fs.StringVar(&o.buildCluster, "build-cluster", "", "Path to kube.Cluster YAML file. If empty, uses the local cluster.")
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply prow/flagutil/kubernetes_cluster_clients.go]: They technically don't have to be, anymore

File: prow/flagutil/kubernetes_cluster_clients.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	if o.kubeconfig != "" && o.buildCluster != "" {
+		return errors.New("must provide only --build-cluster OR --kubeconfig")
+	}
+
+	return nil
+}
+
+// resolve loads all of the clients we need and caches them for future calls.
+func (o *KubernetesOptions) resolve(dryRun bool) (err error) {
+	if o.resolved {
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply prow/flagutil/kubernetes_cluster_clients.go]: I could, if you'd like

File: prow/flagutil/kubernetes_cluster_clients.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		if err != nil {
+			return err
+		}
+		clients[context] = client
+	}
+
+	if o.infraContext == "" {
+		o.infraContext = defaultContext
+	}
+	infraConfig := clusterConfigs[o.infraContext]
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply prow/flagutil/kubernetes_cluster_clients.go]: Nice catch

File: prow/flagutil/kubernetes_cluster_clients.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	dryRun                     bool
+	prowJobClientset           prow.Interface
+	kubernetesClientsByContext map[string]kubernetes.Interface
+}
+
+// AddFlags injects Kubernetes options into the given FlagSet.
+func (o *KubernetesOptions) AddFlags(fs *flag.FlagSet) {
+	fs.StringVar(&o.buildCluster, "build-cluster", "", "Path to kube.Cluster YAML file. If empty, uses the local cluster.")
+	fs.StringVar(&o.kubeconfig, "kubeconfig", "", "Path to .kube/config file. If empty, uses the local cluster.")
+	fs.StringVar(&o.infraContext, "context", "", "The name of the kubeconfig context to use for the infrastructure client. If empty, uses the local cluster.")
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply prow/flagutil/kubernetes_cluster_clients.go]: ACK

File: prow/flagutil/kubernetes_cluster_clients.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+	resolved                   bool
+	dryRun                     bool
+	prowJobClientset           prow.Interface
+	kubernetesClientsByContext map[string]kubernetes.Interface
+}
+
+// AddFlags injects Kubernetes options into the given FlagSet.
+func (o *KubernetesOptions) AddFlags(fs *flag.FlagSet) {
+	fs.StringVar(&o.buildCluster, "build-cluster", "", "Path to kube.Cluster YAML file. If empty, uses the local cluster.")
+	fs.StringVar(&o.kubeconfig, "kubeconfig", "", "Path to .kube/config file. If empty, uses the local cluster.")
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply prow/flagutil/kubernetes_cluster_clients.go]: ACK

**Comment by stevekuznetsov** [comment]
> IMO it'd be simpler to switch everything at once, if that is the intention... otherwise I have to change the deployment manifests N times (with each prow bump)

@munnerz I did this but it's like a 2K LOC change due to tests and I figured I've landed enough of those in the last week ... so I have all the work done and will open all the PRs soon

**Comment by stevekuznetsov** [comment]
@munnerz updated

File: prow/kube/dry_run_client.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (fejta) [inline prow/kube/dry_run_client.go]: Wrong year?

**Comment by stevekuznetsov** [comment]
@fejta updated all dates to be 2019

**Comment by stevekuznetsov** [comment]
Looks like we missed that webhook

/retest

**Pull request merged**